DOCSP-48383 document bson (#1044)
mongodb/docs-node: source/data-formats/bson.txt

**BSON**, or Binary JSON, is the data format that MongoDB uses to organize
and store data. This data format includes all JSON data structure types and
adds support for types including dates, different size integers, ObjectIds, and
binary data. You can use BSON documents in your {+language+} application by importing the
BSON package. For a complete list of supported types, see the
:manual:`BSON Types </reference/bson-types>` server manual page.

The code samples in this guide use the following BSON document as an example:

.. code-block:: none

   {
       "address" : {
           "street" : "Pizza St", 
           "zipcode" : "10003" 
       },
       "coord" : [-73.982419, 41.579505],
       "cuisine" : "Pizza",
       "name" : "Mongo's Pizza"
   }

Create a BSON Document
----------------------

You can create a BSON document by using the same notation you use to create an
object in {+language+}. The {+driver-short+} automatically converts {+language+} objects
into BSON documents when inserting them into a collection.

The following example creates a BSON document that
represents the preceding sample BSON document:

.. code-block:: javascript

>    const document = {

Reviewer: Maybe my Node is just weak, but isn't this immutable? This may be a question for tech review, but there must me a more general way to create a BSON-structure object.

Author: I'm a little confused at what you mean by immutable in this case! When I tested this code I was able to change the document created using the code in the "change a bson document" section of this doc, and to my knowledge generally objects are mutable in Node JS. I also don't know much about Node or JS, but I was basing this doc off the [PyMongo BSON Doc](https://www.mongodb.com/docs/languages/python/pymongo-driver/current/data-formats/bson/) and this [Stack Overflow thread](https://stackoverflow.com/questions/25977832/how-do-i-convert-a-javascript-object-to-a-mongodb-bson-object).

Reviewer 2: `const` here only means the variable cannot be reassigned, the object contents is totally mutable :)

       "address": {
           "street": "Pizza St",
           "zipcode": "10003",
         },
       "coord": [-73.982419, 41.579505],
       "cuisine": "Pizza",
       "name": "Mongo's Pizza",
   }

Change a BSON Document
----------------------

You can modify the contents of a BSON document by using the same notation you use to modify
an object in {+language+}. The following example makes three changes to the previous
BSON document:

1. Adds a new field, ``restaurant_id``, with the value ``12345``
#. Removes the ``cuisine`` field
#. Sets the value of the ``name`` field to ``"Mongo's Pizza Place"``

.. code-block:: javascript

   document.restaurant_id = "12345";
   delete document.cuisine;
   document.name = "Mongo's Pizza Place";

Write BSON to a File
--------------------

To write BSON data to a file, import the file system module and open the output file.
Then, write each document to the output file. Ensure that documents are encoded in BSON
format by using the ``BSON.serialize()`` method.

The following example writes the sample BSON document to ``file.bson``:

.. code-block:: javascript

>    const fs = require('fs');  // Import the file system module
>    const BSON = require('bson');  // Import the BSON package

Reviewer 2: ```suggestion
   import fs from 'fs/promises';  // Import the file system module
   import { BSON } from 'bson';  // Import the BSON package
```

Just using modern practices here

   // Create a BSON object
   const bsonData = BSON.serialize(result);

   // Write the BSON data to a file

>    fs.writeFile('file.bson', bsonData, (err) => {
>       if (err) throw err;
>       console.log('BSON data written to file.bson');
>     });

Reviewer 2: ```suggestion
   await fs.writeFile('file.bson', bsonData);
   console.log('BSON data written to file.bson');
```

Along with the change to use the promises fs API

Read BSON from a File
---------------------

To read BSON documents from a file, open a file in read mode. Then, decode the documents
from BSON format as you read them by using the ``BSON.deserialize()`` method.

The following example reads the sample BSON document from ``file.bson``:

.. io-code-block::
   :copyable: true
     
   .. input::
      :language: javascript

>       const fs = require('fs');  // Import the file system module
>       const BSON = require('bson');  // Import the BSON package

Reviewer 2: ```suggestion
      import fs from 'fs/promises';  // Import the file system module
      import { BSON } from 'bson';  // Import the BSON package
```

      // Read the BSON data from a file

>       fs.readFile('file.bson', (err, data) => {
>          if (err) throw err;
>          const document = BSON.deserialize(data);
>          console.log(document);
>       });

Reviewer 2: ```suggestion
      const data = await fs.readFile('file.bson');
      const document = BSON.deserialize(data);
      console.log(document);
```
(again just promises)

   .. output::
      :visible: false

>       {
>       _id: new ObjectId('67e1823d0d63bfdf87e8928e'), 
>       "address": {"street": "Pizza St", "zipcode": "10003"}, 
>       "coord": [-73.982419, 41.579505], 
>       "cuisine": "Pizza", 
>       "name": "Mongo's Pizza"
>       }

Reviewer 2: ```suggestion
      {
         _id: new ObjectId('67e1823d0d63bfdf87e8928e'),
         address: { street: 'Pizza St', zipcode: '10003' },
         coord: [ -73.982419, 41.579505 ],
         cuisine: 'Pizza',
         name: "Mongo's Pizza"
      }
```

If we want to match this up with what users will see the keys will most likely be unquoted, which I think is worth showing to draw the distinction between plain JSON and javascript code more clearly.
